src/garage/torch/algos/ddpg.py

```diff
@@ -0,0 +1,300 @@
+                    tabular.record('AverageSuccessRate',
+                                   np.mean(self.success_history))
+
+            if not self.smooth_return:
+                self.episode_rewards = []
+                self.episode_policy_losses = []
+                self.episode_qf_losses = []
+                self.epoch_ys = []
+                self.epoch_qs = []
+
+            self.success_history.clear()
+
+        return last_average_return
+
+    def optimize_policy(self, itr, samples_data):
+        """
+        Perform algorithm optimizing.
+
+        Returns:
+            action_loss: Loss of action predicted by the policy network.
+            qval_loss: Loss of Q-value predicted by the Q-network.
+            ys: y_s.
+            qval: Q-value predicted by the Q-network.
+
+        """
+        transitions = self.replay_buffer.sample(self.buffer_batch_size)
+        transitions = self.np_to_torch(transitions)
+        observations = transitions['observation']
+        rewards = transitions['reward']
+        actions = transitions['action']
+        next_observations = transitions['next_observation']
+        terminals = transitions['terminal']
+
+        rewards = rewards.reshape(-1, 1)
+        terminals = terminals.reshape(-1, 1)
+
+        if self.input_include_goal:
+            goals = transitions['goal']
+            next_inputs = np.concatenate((next_observations, goals), axis=-1)
+            inputs = np.concatenate((observations, goals), axis=-1)
+        else:
+            next_inputs = next_observations
+            inputs = observations
+        with torch.no_grad():
+            next_actions = self.target_policy._nn_module(next_inputs)
+            target_qvals = self.target_qf.get_qval(next_inputs, next_actions)
+
+        clip_range = (-self.clip_return,
+                      0. if self.clip_pos_returns else self.clip_return)
+
+        y_target = rewards + (1.0 - terminals) * self.discount * target_qvals
+        y_target = torch.clamp(y_target, clip_range[0], clip_range[1])
+
+        # optimize critic
+        qval = self.qf.get_qval(inputs, actions)
+        qval_loss = self.qf_criterion(qval, y_target)
+        self.qf_optimizer.zero_grad()
+        qval_loss.backward()
+        self.qf_optimizer.step()
+
+        # optimize actor
+        actions = self.policy._nn_module(inputs)
+        action_loss = -1 * self.qf.get_qval(inputs, actions).mean()
+        self.policy_optimizer.zero_grad()
+        action_loss.backward()
+        self.policy_optimizer.step()
+
+        # update target networks
+        self.update_target()
+        return (qval_loss.detach(), y_target, qval.detach(),
+                action_loss.detach())
+
+    def update_target(self):
+        """Update parameters in the target policy and Q-value network."""
+        for t_param, param in zip(self.target_qf._nn_module.parameters(),
+                                  self.qf._nn_module.parameters()):
```

Comment:
The fact that we have to reach into the q function like this reinforces my view that the q function should just be an appropriately shaped neural network. This doesn't need to be addressed in this change, but it would be good to fix soon.